fabric-transfer-api-v1/src/main/java/net/fabricmc/fabric/api/transfer/v1/context/ContainerItemContext.java

```diff
@@ -0,0 +1,161 @@
+ * For example, it allows a water bucket to replace itself by an empty bucket when emptied.
+ * A {@code ContainerItemContext} generally represents a slot in an inventory, and additionally allows access to the rest of the inventory,
+ * which may or may not contain the preferred slot itself.
+ *
+ * <p>When an {@linkplain ItemApiLookup item API} requires a {@code ContainerItemContext} as context,
+ * it will generally be suitable to obtain a context instance with {@link #ofPlayerHand} or {@link #ofPlayerCursor},
+ * and then use {@link #find} to query an API instance.
+ *
+ * <p>Implementors of item APIs will generally want to:
+ * <ul>
+ *     <li>Query how much of the variant is in the inventory through {@link #getAmount}.</li>
+ *     <li>Transform some of the bound variant into another variant through {@link #transform}.</li>
+ * </ul>
+ */
+// TODO: update javadoc
+// TODO: wrapper around a SingleSlotStorage<ItemVariant>
+// TODO: wrapper around a StackReference maybe?
+public interface ContainerItemContext {
+	/**
+	 * Try to find an API instance for the passed lookup and return it, or {@code null} if there is none.
+	 * The API is queried for the current variant, if it's not blank.
+	 *
+	 * @see ItemApiLookup#find
+	 */
+	@Nullable
+	default <A> A find(ItemApiLookup<A, ContainerItemContext> lookup) {
+		return getItemVariant().isBlank() ? null : lookup.find(getItemVariant().toStack(), this);
+	}
+
+	/**
+	 * Return the current item variant of this context, that is the variant in the slot of the context.
+	 * If the result is non blank, {@link #getAmount} should be
+	 */
+	ItemVariant getItemVariant();
+
+	/**
+	 * Return the current amount of {@link #getItemVariant()} in the slot of the context.
+	 *
+	 * @throws IllegalStateException If {@linkplain #getItemVariant() the current variant} is blank.
+	 */
+	long getAmount();
+
+	/**
+	 * Try to insert some items into this context, prioritizing the slot of the context over the rest of the inventory.
+	 *
+	 * @see Storage#insert
+	 */
+	long insert(ItemVariant itemVariant, long maxAmount, TransactionContext transaction);
+
+	/**
+	 * Try to extract some items from this context's slot.
+	 *
+	 * @see Storage#extract
+	 */
+	long extract(ItemVariant itemVariant, long maxAmount, TransactionContext transaction);
+
+	/**
+	 * Try to transform as many as possibly of {@linkplain #getItemVariant() the current variant} into another variant.
+	 * That is, extract the old variant, and insert the same amount of the new variant instead.
+	 *
+	 * @param into The variant of the items after the conversion. May not be blank.
+	 * @param maxAmount The maximum amount of items to convert. May not be negative.
+	 * @param transaction The transaction this operation is part of.
+	 * @return A nonnegative integer not greater than maxAmount: the amount that was transformed.
+	 */
+	default long transform(ItemVariant into, long maxAmount, TransactionContext transaction) {
```

Comment:
I would change this into 
`long transform(ItemVariant from, long amount, ItemVariant to, long amount, TransactionContext ctx)`
For example, if an item were to drain water from buckets in the inventory, it may be desirable to keep all the buckets in the same positions in the inventory, in which case it could be called once for each bucket stack, but if stacking is desired, then the function can be called once with the total remainder of buckets.

Reply:
In-place modification can already be done through the slot list.